rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,859 @@
+This is to prevent other [`Focusable`]s from stealing focus while dragging
+around the knob with the mouse.
+
+We then use mouse movement to move around the knob,
+and update the audio level based on the knob's position.
+
+We send a `NavRequest::Free` when the player release's the left mouse button.
+
+
+### Custom directional navigation
+
+Beyond just changing the inputs that generate [`NavRequest`]s,
+it's possible to customize gamepad-style directional input generated by
+`NavRequest::Direction`.
+The `NavigationPlugin` is generic over directional input.
+It accepts a `M: MenuNavigationStrategy` type parameter.
+
+The default implementation, the one for `bevy_ui`, uses `GlobalTransform`
+to resolve position.
+
+See [relevant implementation section](#MenuNavigationStrategy).
+
+
+## Implementation
+
+The nitty-gritty code… is already available!
+The [`bevy-ui-navigation`] repo contains the code.
+[this is a good start for people interested in the architecture][ui-nav-arch].
+
+### Internal representation
+
+The navigation tree is a set of relationships between entities with either the
+[`Focusable`] component or the [`TreeMenu`] component.
+
+The hierarchy is specified by both the native bevy hierarchy
+and the `focus_parent` of the [`TreeMenu`] component.
+
+In the following screenshot, `Focusable`s are represented as circles, menus
+as rectangles, and the `focus_parent` by blue arrows
+(menu points to its parent).
+
+![A screenshot of a RPG-style menu with the navigation tree overlayed](https://user-images.githubusercontent.com/26321040/141671969-ea4a461d-0949-4b98-b060-c492d7b896bd.png)
+
+Note that gold buttons are `FocusState::Active` while the orange-red button
+("B") is `FocusState::Focused`.
+
+Since the "tabs menu" doesn't have an outgoing arrow, it is the root menu.
+
+The active buttons are a breadcrumb of buttons to activate to reach
+the current menu with the `Focused` focusable from the root menu.
+
+To move the focus from the "soul menu" to the "ABC menu",
+you need to send `NavRequest::Action` while the button "abc" is focused
+(i.e.: _"activating"_ the button).
+
+Such a move would generate the following `NavEvent`:
+```rust
+NavEvent::FocusChanged {
+  to: [<B entity>, <abc entity>],
+  from: [<abc entity>],
+}
+```
+
+If there were a "KFC menu" (currently hidden) [child of](#Terminology)
+the "kfc" button, then [activating](#Terminology)
+the "kfc" button would send the focus to
+the prioritized focusable within the "KFC menu".
+
+To navigate from "ABC menu" back to "soul menu",
+you would send `NavRequest::Cancel`.
+Such a move would generate the following `NavEvent`:
+```rust
+NavEvent::FocusChanged {
+  to: [<abc entity>],
+  from: [<B entity>, <abc entity>],
+}
+```
+The "tabs menu" is defined as a "scope menu", which means that
+by default, the `LT` and `RT` gamepad buttons will navigate
+the "tabs menu" regardless of the current focus position.
```

Comment:
I'm guessing this is just for illustrative purposes, but we should aim to separate button inputs from action events, which would be out of scope for this RFC (and I suspect @alice-i-cecile's input manager is making a good job of that) but could this be example be rewritten to use something like `NavRequest::Next` and `NavRequest::Previous` (as suggested above)?